File: cplus/leetcode/202512/algo1808.cpp
#include <bits/stdc++.h>
using namespace std;

// 杂项 - 7.2 整数拆分
class Solution {
public:
    int maxNiceDivisors(int primeFactors) {
        const int mod = 1e9 + 7;
        if (primeFactors < 5) return primeFactors;
        
        auto qpow = [] (long long x, int e) -> long long {
            long long res = 1;
            while (e) {
                if (e & 1) res = res * x % mod;

                e >>= 1;
                x = x * x % mod;
            }
            return res;
        };

        int expo = primeFactors / 3, rem = primeFactors % 3;
        if (rem == 0) rem = 1;
        else if (rem == 1) expo --, rem += 3;
        return qpow(3, expo) * rem % mod; 
    }
};

File: python/algo/202512/algo1808.py
# 杂项 - 7.2 整数拆分

class Solution:
    def maxNiceDivisors(self, primeFactors: int) -> int:
        mod = 10**9 + 7
        ans = 1
        # 遍历 k 从 2 到 primeFactors（包含）
        if primeFactors < 5: return primeFactors
        expo = primeFactors // 3  # 整除，对应 C++ 的 /（int 除法）
        rem = primeFactors % 3    # 取余
        if rem == 0: rem = 1
        elif rem == 1:
            expo -= 1
            rem = 4
        res = pow(3, expo, mod) * rem % mod

        # 返回结果取模
        return res      